File: Ejemplos/AVL.cpp
#include <iostream>
#include <string>

using namespace std;

class nodo{
    public:
        nodo *izq, *der;
        int dato;
        int altura;

        nodo(int dat){
            this->dato = dat;
            izq = der  = nullptr;
            altura = 0;
        }

};

class AVL{
    private:
        int MAX(int val1, int val2){
            if(val1 > val2)
                return val1;
            return val2;
        }
        nodo * RotacionIzquierda(nodo *pivote){
            nodo *aux = pivote->izq;
            pivote->izq = aux->der;
            aux->der = pivote;
            pivote->altura = MAX(getAltura(pivote->izq), getAltura(pivote->der))+1;
            aux->altura = MAX(getAltura(aux->izq),pivote->altura)+1;
            return aux;
        }
        nodo * RotacionDerecha(nodo *pivote){
            nodo *aux = pivote->der;
            pivote->der = aux->izq;
            aux->izq = pivote;
            pivote->altura = MAX(getAltura(pivote->der), getAltura(pivote->izq))+1;
            aux->altura = MAX(getAltura(aux->der),pivote->altura)+1;
            return aux;
        }

        nodo * RotacionDobleIzquierda(nodo *pivote){
            pivote->izq = RotacionDerecha(pivote->izq);
            return RotacionIzquierda(pivote);
        }
        nodo *RotacionDobleDerecha(nodo *pivote){
            pivote->der = RotacionIzquierda(pivote->der);
            return RotacionDerecha(pivote);
        }

        int getAltura(nodo *temp){
            if(temp!=nullptr){
                return temp->altura;
            }
            return -1;
        }

        nodo * insert(nodo *root, int dato){
            if(root==nullptr){
                root = new nodo(dato);
            }else if(dato<root->dato){
                root->izq = insert(root->izq,dato);
                if(getAltura(root->der)-getAltura(root->izq)==-2){
                    if(dato<root->izq->dato){
                        root = RotacionIzquierda(root);
                    }else{
                        root = RotacionDobleIzquierda(root);
                    }
                }
            }else if(dato> root->dato){
                
                root->der = insert(root->der,dato);
                if(getAltura(root->der)-getAltura(root->izq)==2){
                    if(dato>root->der->dato){
                        root = RotacionDerecha(root);
                    }else{       
                        root = RotacionDobleDerecha(root);
                    }
                }
            }
            root->altura = MAX(getAltura(root->izq),getAltura(root->der))+1;
            return root;

        }
        void recorridoInOrden(nodo *pivote){
            if(pivote!=nullptr){
                recorridoInOrden(pivote->izq);
                cout<< pivote->dato<< " ";
                recorridoInOrden(pivote->der);
            }
            return;
        }

    public: 
        nodo *root;
        AVL(){
            root = nullptr;
        }
        void insertar(int dato){
           root = insert(root, dato);
        }
        void print(){
            recorridoInOrden(root);
        }

};

int main(){
    AVL *avl = new AVL();
    avl->insertar(1);
    avl->insertar(2);
    avl->insertar(3);
    avl->insertar(4);
    avl->insertar(5);
    avl->insertar(6);
    avl->print();
    return 0;
}

File: Go/avl/avl.go
package avl

import "fmt"

type nodo struct {
	indice   int
	altura   int
	izq, der *nodo
}

func newNodo(indice int) *nodo {
	return &nodo{indice, 0, nil, nil}
}

type AVL struct {
	raiz *nodo
}

func NewAVL() *AVL {
	return &AVL{nil}
}

func max(val1 int, val2 int) int {
	if val1 > val2 {
		return val1
	}
	return val2
}

func altura(temp *nodo) int {
	if temp != nil {
		return temp.altura
	}
	return -1
}

func rotacionIzquierda(temp **nodo) {
	aux := (*temp).izq
	(*temp).izq = aux.der
	aux.der = *temp
	(*temp).altura = max(altura((*temp).der), altura((*temp).izq)) + 1
	aux.altura = max(altura((*temp).izq), (*temp).altura) + 1
	*temp = aux
}

func rotacionDerecha(temp **nodo) {
	aux := (*temp).der
	(*temp).der = aux.izq
	aux.izq = *temp
	(*temp).altura = max(altura((*temp).der), altura((*temp).izq)) + 1
	aux.altura = max(altura((*temp).der), (*temp).altura) + 1
	*temp = aux
}

func rotacionDobleIzquierda(temp **nodo) {
	rotacionDerecha(&(*temp).izq)
	rotacionIzquierda(temp)
}

func rotacionDobleDerecha(temp **nodo) {
	rotacionIzquierda(&(*temp).der)
	rotacionDerecha(temp)
}

func insert(indice int, root **nodo) {
	if (*root) == nil {
		*root = newNodo(indice)
		return
	}
	if indice < (*root).indice {
		insert(indice, &(*root).izq)
		if (altura((*root).izq) - altura((*root).der)) == -2 {
			if indice < (*root).izq.indice{
				rotacionIzquierda(root)
			}else{
				rotacionDobleIzquierda(root)
			}
		}
	}else if indice > (*root).indice{
		insert(indice, &(*root).der)
		if (altura((*root).der) - altura((*root).izq)) == 2{
			if indice > (*root).der.indice {
				rotacionDerecha(root)
			}else{
				rotacionDobleDerecha(root)
			}
		}
	}else{
		fmt.Println("Ya se inserto el indice")
	}

	(*root).altura = max(altura((*root).izq), altura((*root).der))+1
}

func (avl *AVL) Insertar(indice int) {
	insert(indice, &avl.raiz)
}

func (avl *AVL) Print(){
	inOrden(avl.raiz)
}

func inOrden(temp *nodo)  {
	if temp != nil {
		inOrden(temp.izq)
		fmt.Println("Index: ", temp.indice)
		inOrden(temp.der)
	}
}
